### ISSUE TITLE
[BUG] Ausgov (AU-5) blocks requests, with a 403 status code response

### ISSUE THREAD

**Issue by andylolz** [codeforIATI/iati-data-bugtracker#9]
### Explanation of the bug

Since around July 2020, ausgov (AU-5) servers hosting IATI data appear to have been responding with a 403 status code to various services that consume IATI data. I suspect they are blacklisting these services by IP address, but that’s unclear.

When I request the data from my machine, it works fine, e.g. this dataset:
https://iatiregistry.org/dataset/ausgov-af

However, we can see d-portal has had trouble downloading the same dataset:
http://d-portal.org/ctrack.html#view=dash_sluglog&slug=ausgov-af
![Screenshot 2021-01-19 at 15 50 51](https://user-images.githubusercontent.com/464193/105058447-1fb83380-5a6e-11eb-8dbe-87bffbdc94f4.png)

In fact, d-portal hasn’t successfully downloaded any ausgov data since their recent server reboot:
http://d-portal.org/ctrack.html#view=search&reporting_ref=AU-5
![Screenshot 2021-01-19 at 15 50 25](https://user-images.githubusercontent.com/464193/105058385-1038ea80-5a6e-11eb-8b43-fbadc0894dd9.png)

The new datastore also seems to have had problems with it:
https://iatidatastore.iatistandard.org/search/activity?q=reporting_org_ref:(AU-5)&wt=json&rows=50
```json
{
  …
  "response": {
    "numFound": 0,
    "start": 0,
    "docs": []
  }
}
```

…And IATI data dump is showing errors for all ausgov data:
https://gist.githubusercontent.com/codeforIATIbot/f117c9be138aa94c9762d57affc51a64/raw/e9e26621d812b89789c6bbc8697fb4461bbb974e/errors

**Comment by andylolz** [comment]
### Does the publisher know?

I contacted them today.

**Comment by notshi** [comment]
Also raised here https://github.com/zimmerman-team/iati.cloud/issues/2470

**Comment by andylolz** [comment]
Thanks, @notshi! I’ve added a comment there.

**Comment by andylolz** [comment]
It’s definitely still applicable.

**Comment by andylolz** [comment]
> would you mind checking to see if the issue is still applicable?

**It is, yes.** I suspect some services have been whitelisted on Australia DFAT servers, but others have not.

The issue was raised on the registry (https://github.com/IATI/ckanext-iati/issues/323), but after changing the `user-agent` header (https://github.com/IATI/ckanext-iati/commit/d195e2deed4d6ef77487b9ceb7fa0b2b75bbce49), it looks like **the registry archiver is successfully fetching these datasets**.

Also, **the new datastore now appears to successfully fetch these datasets**:
https://iatidatastore.iatistandard.org/search/activity?q=reporting_org_ref:(AU-5)&wt=json&rows=50

```json
{
  …
  "response": {
    "numFound": 13632,
    "start": 0,
    "docs": […]
  }
}
```

**However**, d-portal is still having trouble downloading Australia DFAT datasets:
http://d-portal.org/ctrack.html#view=dash_sluglog&slug=ausgov-af
![Screenshot 2021-05-01 at 12 15 28](https://user-images.githubusercontent.com/464193/116780844-f2205480-aa76-11eb-82a2-c8b6174e7ef3.png)

http://d-portal.org/ctrack.html#view=search&reporting_ref=AU-5
![Screenshot 2021-05-01 at 13 00 04](https://user-images.githubusercontent.com/464193/116781784-2c8cf000-aa7d-11eb-83d9-ae2d71c9a646.png)

The old datastore no longer exists, but datastore classic has never managed to fetch Australia DFAT data. E.g.:
https://datastore.codeforiati.org/api/1/about/dataset/ausgov-af

```json
{
    "dataset": "ausgov-af",
    "last_modified": "2021-03-31T23:15:12.448867",
    "num_resources": 1,
    "resources": [
        {
            "last_fetch": "2021-03-02T00:40:31",
            "last_parsed": null,
            "last_status_code": 404,
            "last_successful_fetch": null,
            "num_of_activities": 0,
            "url": "https://www.dfat.gov.au/sites/default/files/Australian_Aid_Country_File_Afghanistan.xml"
        }
    ]
}
```

(NB the `last_status_code` is wrong here – that’s [a known bug](https://github.com/codeforIATI/iati-datastore/issues/252).)

**Comment by notshi** [comment]
Thanks for the update, @andylolz 

d-portal downloads the files randomly and in no particular order so assuming we are allowed to at least get one file a day, we might eventually get all of them. However, as you've mentioned, **we are probably not whitelisted**.

Via our [database logs](https://xriss.github.io/D-Portal-Logs/xpath.html?pid=AU-5#/iati-activities/iati-activity/iati-identifier), the last last successful fetch was 15 Dec 2020 for 13,650 activities.

This was the same day the d-portal server died due to a disk error which took a week to be replaced and installed, and our ip address would have also changed as a result. When the hard disk on the server failed, we lost any data that was currently in a cached state. **This would explain why there is no 'last successfully downloaded' date for AU-5 datasets on Dash.**

The Dashboard is also having trouble downloading the files.
http://dashboard.iatistandard.org/publisher/ausgov.html
![Screenshot-20210501142412-573x586](https://user-images.githubusercontent.com/1515961/116783984-a4611780-aa89-11eb-83a3-b5365194ad28.png)

Looks like it's been happening since [2020-07-13 19:36:57 +0100](http://dashboard.iatistandard.org/stats/gitaggregate-publisher-dated/ausgov/activities.json).

Oddly, there was a blip on 2021-04-21 14:52:12 +0100 where the Dashboard was able to retrieve 13,927 activities!

**Comment by matmaxgeds** [comment]
I am not sure it feels 'right', but where the DSv2 has access, could we set it as a backup source - I think they have an internal url listed for each file: https://iatidatastore.iatistandard.org/api/datasets/?publisher_identifier=AU-5&format=json but not sure if it actually accessible

**Comment by andylolz** [comment]
> where the DSv2 has access, could we set it as a backup source - I think they have an internal url listed for each file

Interesting – those `internal_url`s appear to have saved the HTML of a 404 page, [that looks like this](https://lyj9k.csb.app). That possibly means DSv2 is also struggling to import Australia DFAT data.

**Comment by andylolz** [comment]
> where the DSv2 has access, could we set it as a backup source - I think they have an internal url listed for each file

Oooh – this is useful to know about, thanks! Interesting – those `internal_url`s appear to have saved the HTML of a 404 page, [that looks like this](https://lyj9k.csb.app). That possibly means DSv2 is also struggling to import Australia DFAT data.

**Comment by andylolz** [comment]
Thanks for this, @notshi!

> The Dashboard is also having trouble downloading the files.
> http://dashboard.iatistandard.org/publisher/ausgov.html

I did not think to check the dashboard! The codeforIATI version doesn’t list ausgov as a publisher at all:
https://dashboard.codeforiati.org/publisher/ausgov.html

**Comment by andylolz** [comment]
I moved [iati-data-dump](https://iati-data-dump.codeforiati.org) to github actions today, and it successfully managed to download ausgov data.

This data should start to bubble up through codeforIATI services, e.g. to [datastore classic](https://datastore.codeforiati.org/) and [the dashboard](https://dashboard.codeforiati.org/).

**Comment by notshi** [comment]
We are using **User-Agent Mozilla/5.0** and that looks to be blocked by AU-5.
This is a problem as we starting using this because other servers block curl.

So now we use the **default curl ua as the the backup**.
And this seems to have solved issues with *many* servers that were previously giving us errors.

We now have over a million activities so it looks like this might have found us 40,000 activities.

http://d-portal.org/ctrack.html#view=search&reporting_ref=AU-5
![Screenshot-20210507124946-524x264](https://user-images.githubusercontent.com/1515961/117445577-c1d32d00-af32-11eb-84a1-3421a80dbfc4.png)

http://d-portal.org/ctrack.html#view=dash_sluglog&slug=ausgov-af
![Screenshot-20210507124904-830x446](https://user-images.githubusercontent.com/1515961/117445580-c26bc380-af32-11eb-9cef-9f9f25e6b366.png)

By the way, there are still **issues** with [some datasets](https://iatiregistry.org/publisher/ausgov?q=error&sort=metadata_modified+desc) by AU-5.

![Screenshot-20210507130441-717x489](https://user-images.githubusercontent.com/1515961/117446987-cf89b200-af34-11eb-89f0-e52d07c30a2f.png)

**Comment by andylolz** [comment]
> So now we use the default curl ua as the the backup.
> And this seems to have solved issues with many servers that were previously giving us errors.

I’m still unsure how this blocking works… Perhaps a combination of `user-agent` and IP address? (I didn’t change `user-agent` but did change IP address, and that also did the trick).

> By the way, there are still issues with some datasets by AU-5.

That’s true… But those are 404 errors, so I think we should consider them separately.

**Comment by andylolz** [comment]
I’m not really sure what to do with this ticket? I’m not convinced it’s fixed, but it sounds like we’re not experiencing the problem right now…

**Comment by andylolz** [comment]
> I think the ticket should still be opened due to ongoing issues

Cool okay, agreed.

> also because the Dashboard is still getting errors accessing the data

Oh, good point – you’re right.

**Comment by notshi** [comment]
Ok so this might trickle down to who/what tools we are adding data issues for.

Might be worth updating the readme?

So for example, this repo tracks issues that affect externally maintained tools using IATI, etc. This should include the Registry.

I mean, I'd like to consider this issue closed but AU-5 might hiccup tomorrow or in a week's time. Though we could re-open or create a new issue if that happens.

**Comment by notshi** [comment]
By the way, feel free to ignore suggestions if they seem pedantic! It's mostly for my train of thought and process.

**Comment by andylolz** [comment]
No problem at all!

Let’s make a new `meta` ticket to decide what should/shouldn’t be recorded.

**Comment by siemvaessen** [comment]
Looks available? IATI.cloud had some issues on this as well, but https://iatidatastore.iatistandard.org/search/activity?q=reporting_org_ref:(AU-5)&wt=json&rows=13632 seems to show all AU-5 data?

**Comment by andylolz** [comment]
> Looks available? IATI.cloud had some issues on this as well, but https://iatidatastore.iatistandard.org/search/activity?q=reporting_org_ref:(AU-5)&wt=json&rows=13632 seems to show all AU-5 data?

The problem is, the DFAT server is a bit over-zealous in its blocking. So while lots of requests succeed (that link works for me, too) a lot of requests are getting blocked. Limiting availability is probably bad practice when serving open data.

For instance, [the IATI dashboard still appears to be unable to access DFAT data](http://dashboard.iatistandard.org/publisher/ausgov.html):
![Screenshot 2021-06-01 at 18 12 15](https://user-images.githubusercontent.com/464193/120364240-e7bada00-c304-11eb-86c0-633f89706720.png)

**Comment by siemvaessen** [comment]
Yeah, we're aware of this issue. We also decided to not spend any more time on this as the data owner is seemingly very reluctant to make any changes on their end.

**Comment by codeforIATIbot** [comment]
Hello! There has been no activity on this issue in the last 30 days. I wonder if it has now been resolved?

If you’re reading this, would you mind checking to see if the issue is still applicable?

* If it has been resolved, please add a message to that effect, and (if you’re able to) close it.
* If the problem still applies, please add a message to let us know. I’ll remove the "awaiting update" label, and will check in again next month.

Thank you!

**Comment by notshi** [comment]
The Dashboard is still unable to access DFAT (AU-5) data / servers.

**Comment by andylolz** [comment]
> In such cases, should there be a new label where a bug is defined as "won't fix" and no reminder set? @andylolz

^^ Yeah, there’s an `evergreen` label for this purpose.
